:book: :sparkles: :people_holding_hands: add proposal for Node Bootstrapping working group

t-lo opened the pull request:
**What this PR does / why we need it**:

Propose a working group for node bootstrapping and cluster provisioning.
The need for this working group originated from an ongoing discussion around separating cluster provisioning and node bootstrapping, as stated in the WG's User Story.

**Which issue(s) this PR fixes**

- https://github.com/kubernetes-sigs/cluster-api/issues/5294
- https://github.com/kubernetes-sigs/cluster-api/issues/9157

**CC**

- "@elmiko"
- "@eljohnson92"
- "@johananl"
- "@tormath1"

**Tags**

/area provider/bootstrap-kubeadm
/area bootstrap

/kind documentation
/kind proposal

t-lo commented:
"@elmiko" "@eljohnson92" I took the liberty to add you as co-stakeholders to the WG proposal - IIRC you expressed interest in participating. I hope that's OK?

johananl commented on docs/community/20241112-node-provisioning.md:
```diff
@@ -0,0 +1,88 @@
+---
+title: ClusterAPI Node Provisioning working group
```
I'm a bit concerned about carrying over the conflation between bootstrap and provisioning into the working group...
The filename has "node provisioning" in it but the title says "node bootstrapping".

Also, I'm not sure what we mean by "cluster provisioning".

To clarify, the two concerns which are currently intertwined are the following:

1. Generating cloud-config or Ignition configuration and executing it on nodes to make OS customizations, file manipulations on disk etc.
2. Generating kubeadm configuration and executing kubeadm to turn a host into a k8s node.

I don't care too much what we call each concern as long as the two are clearly demarcated and we know which one we're discussing at any given moment.

In my WIP design document I offered to call no. 1 "provisioning" and no. 2 "bootstrap".

t-lo replied:
Good catch, I was trying to use "cluster provisioning" and "node bootstrapping" consistently but apparently it escaped me in the title.

johananl commented on docs/community/20241112-node-provisioning.md:
```diff
@@ -0,0 +1,90 @@
+Node Bootstrapping.
+
+Provisioning, in this context, refers to configuring and customising the
+node operating system to prepare it to serve as a ClusterAPI worker cluster node.
```
This phrasing is better IMO, thanks for changing.

```suggestion
node operating system to prepare it to serve as a ClusterAPI cluster node.
```

It's not just worker nodes, though :point_up:

fabriziopandini commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,90 @@
+node bootstrapping process and provide specifications on ClusterAPI's expectations
+and requirements for provisioning-time configuration and customisation.
+
+Initial focus of the working group is on separating node bootstrapping and node
+OS configuration / customisation ("node provisioning" from here on).
+At the time of writing, bootstrapping and provisioning are architecturally
+intertwined; both are implemented in the bootstrap provider.
+This increases maintenance effort and effectively prevents re-using provisioning
+features across bootstrap providers.
```
I think that this statement is somehow more hinting a solution (separation) than a problem.

Base on my understanding, we are trying to solve two problems:
- to offer to the users more combinations of provisioner/bootstrappers.
- to ensure those combinations can evolve freely, taking befits from the greatest an the latest the underlying technologies

But IMO, users don't care about separation at all. 
In fact, users will always need both a provisioner and a bootstrapper to get a working cluster.

If instead we split provisioner and a bootstrapper into two separate components, probably we are not making a good service both to the users, nor to ourself (as project maintainers), nor to the community that builds on top of Cluster API:

- we are adding to the system another dimension of complexity, that makes this change highly disruptive for users
- we are imaging a change that will impact not only the users, but the entire of the CAPI ecosystem, think e.g. to the CAPI operator, or all the products built on top of Cluster API.
- we are adding to the users an additional operational burden (they have to take care of yet another component/provider, its upgrades, its compatibility matrix, its CVEs etc etc). 
- we are imaging a change so complex and extensive that, being very pragmatic, the chances that this change might happen drastically drop (e.g. it will impact clusterctl, clusterclass, runtime extensions, all the controllers, almost everything; in fact this is why this problem is laying around for so long)

Frankly speaking, this seems too much to digest for everyone. 

The good news, it that I think that it is possible to look at the problem above from a different angle and find a better solution.

How to solve the two problem above (offer more combinations/ensure those combinations can evolve freely) without introducing separation?

That's up to the working group to figure out, but If I can give my two cents, what we are looking for is a solution that doesn't impact users, but makes it easier to implement and evolve different provisioners (e.g. we want to bump ignition!), while re-using existing bootstrappers (e.g. kubeadm).

I think that the key point in the sentence above is software re-use

So, a possible solution might be in a combination of federated development/separated repositories (providers as we call it in CAPI) + something that makes it easy to reuse code e.g. making kubeadm provider a library (details TBD of course, but I hope you get the idea)

Please also see https://docs.google.com/document/d/1Fz5vWwhWA-d25_QDqep0LWF6ae0DnTqd5-8k8N0vDDM/edit?disco=AAABX57uOnU where I make other considerations about the same point

Happy to chat about it if you are interested on this angle

t-lo replied:
These are good points - will rework the proposal accordingly. Admittedly I wrote the first draft with the existing proposal in mind.

Thank you for your review!

johananl replied:
Thanks @fabriziopandini. I completely understand the concerns you've raised and am likewise worried by them. I'm also 100% for maximizing code reuse.

If I understand correctly, what you are proposing is to keep provisioners (e.g. Ignition) tightly coupled with bootstrappers (e.g. kubeadm) and offer multiple bootstrap providers for each bootstrapper, one for each provisioner. So, for example, we would have a **kubeadm + cloud-config** bootstrap provider and a **kubeadm + Ignition** bootstrap provider in separate repositories.

If that's what you meant, I'm seeing the following problems with this approach:

1. The cartesian product of bootstrappers and provisioners can grow very large. True, with 1 bootstrapper and 2 provisioners we would have only 2 repositories to maintain, but what if we have 3 bootstrappers and 3 provisioners? That's **9 repositories** already. AFAICT this isn't premature optimization since we seem to already have [8 supported bootstrappers](https://cluster-api.sigs.k8s.io/reference/providers.html?highlight=k3s#control-plane) and at least 3 provisioners (cloud-init and Ignition, with CloudBase Init either implemented or planned - not sure). That's 24 repos if we don't separate.

2. Even if we decide the number of repos is manageable, it's not clear to me how we would achieve good code reuse when we have cloud-config fields embedded in types such as `KubeadmConfigSpec`. Example:

```yaml
apiVersion: controlplane.cluster.x-k8s.io/v1beta1
kind: KubeadmControlPlane
metadata:
  name: capi-azure-control-plane
  namespace: default
spec:
  kubeadmConfigSpec:
    ...
    diskSetup:
      filesystems:
      - device: /dev/disk/azure/scsi1/lun0
        extraOpts:
        - -E
        - lazy_itable_init=1,lazy_journal_init=1
        filesystem: ext4
        label: etcd_disk
      ...
      partitions:
      - device: /dev/disk/azure/scsi1/lun0
        layout: true
        overwrite: false
        tableType: gpt
```

In the snippet above, `diskSetup` is a field under `KubeadmConfigSpec`. This field is [specific to cloud-init](https://cloudinit.readthedocs.io/en/latest/reference/modules.html#disk-setup). That means `KubeadmConfigSpec` assumes cloud-init is the only provisioner. True, we could implement "translators" under the hood for Ignition and any other provisioner, but there is no guarantee for feature parity among all provisioners so IMO it's only a matter of time until we run into issues where some feature is supported by cloud-init but not by other provisioners.

I think we have a major design flaw in the current core API (mixing bootstrap and provisioning in the data model). True, it will be very disruptive to change this, but from my understanding you're concerned about building more complexity (e.g. Ignition 3) on top of a suboptimal foundation, and to me it seems we'd be doing exactly that if we keep the API as is and keep adding layers on top of it.

Also, I know that in the past some maintainers have expressed similar concerns: https://github.com/kubernetes-sigs/cluster-api/pull/4172#discussion_r600098540

To summarize, I think that if we avoid tackling the design problem head on and look for ways around it, we'd find ourselves in a similar situation later on, only then we'd be disrupting even more areas of the projects.

I'm open to changing my mind on this :slightly_smiling_face:

fabriziopandini replied:
@johananl that's a complex discussion
you clearly went miles down this path, and everyone else have to catch up (this is what the working group is for).
what I'm asking now is just to be open to other options and to not tight this working group to your work, because this will prevent other's contributions.

Also, please note that, I'm bringing this up mostly because, based on my experience, this could become an issue for the working group down the line when broader consensus will be required.

WRT to the initial assumption of the use cases we have in front of us, I get that theoretically we have to address a possible high number of combinations, but from the other side, as a matter of fact only the ignition team has raised this concern so far.

I'm not sure why, but I think that we should consider that either not all the combinations of providers/bootstrappers makes sense or simply there is no demand for many of those combinations (but getting a better understanding of the problem scope is again part of the working group goal IMO)

WRT to the possible solution, frankly speaking I think it is too early to get into the weed of different options without a better understanding of the context; also this PR is probably not the right place for those discussions.

But just to avoid misunderstanding, I'm well aware of the fact that the current KubeadmConfig API now is not in an ideal state for what we are discussing (as a matter of fact it wasn't designed with re-use in mind). However, what I'm advocating for is to explore options for making re-use possible while minimising the number of breaking changes/impact for users.

But finding a solution for this problem is a discussion that I will be happy to have if the charter of this working group opens up to such options.

johananl replied:
Hey @fabriziopandini. Thanks for your follow up.

Of course I'm open to discussing other solutions and would love to get contributions from others 🙂 I hope I didn't write anything which implies otherwise.

The specific solution in my WIP proposal definitely isn't the only one, and it also isn't a solution I had in mind in advance but rather what I could come up with after gathering a list of current issues and concerns from both users and maintainers around bootstrap and provisioning. I'm not attached to this proposal and am happy to drop it or replace it with a better proposal right away if we decide that's the way to go.

To be clear, the reason I became involved with this work initially was to improve the integration between Flatcar Container Linux and CAPI. So, with my Microsoft hat on I'd be very glad to avoid doing any major reworking of core CAPI pieces if it's not absolutely necessary. We'd be happy to simply keep Ignition supported, add support for Ignition v3 and work towards graduating Ignition support from experimental mode to a stable "GA".

However, I also have the OSS contributor hat, and with that hat on I always try to do what's right for the projects I'm involved in rather than solving ad-hoc issues which stand in my way while working on use cases relevant mainly to my company.

The reason I've gone into the weeds of #5294 and came up with this proposal is that previously we heard very clearly from some of the maintainers that before we can consider making Ignition a fully-supported, "GA" provisioner, we must tackle #5294. If that's no longer the case or if I somehow misunderstood the intention there -- great, as this would make our work around Ignition much shorter and easier 👍 

I'm also in favor of deferring the technical discussions to the work group.

I understand your point (as well as Stefan's) about keeping the API stable. That's very important indeed. The only reason I proposed changing core API objects is that as far as I understood the CAPI maintainers (or at least some of them) required that we cleanly separate bootstrap from provisioning, and I couldn't find a way to truly separate the two without touching the API because the two concerns are unfortunately entangled in the API.

Let's continue this discussion in the first iteration of the work group then 🙂

elmiko replied:
great discussion here, i tend to agree with Fabrizio and Stefan about the need to not disrupt the API and user experience too much. with that said though, i think it would be great if the working group can explore multiple options to see if there is a way we can use the current APIs/contracts, or perhaps add to them instead of modifying, but as noted above this is technical discussion we should have in the WG.

for this proposal, i think we just need to be clear about the goals of the WG and our desire to find solutions that do not require re-architecting core pieces of the project.

t-lo replied:
I've captured our discussion and included deep links to here in the WG agenda doc: https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?usp=sharing

Will close this thread to ease review of this PR as aligned on in the [Nov 27 office hours](https://docs.google.com/document/d/1GgFbaYs-H6J5HSQ6a7n4aKpk0nDLE2hgG2NSOM9YIRw/edit?tab=t.0#heading=h.faoavlb0ym66).

fabriziopandini commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,90 @@
+authors:
+    - "@t-lo"
+reviewers:
+    - "@elmiko"
+    - "@eljohnson92"
+    - "@johananl"
+    - "@tormath1"
```
cc @vincepri @enxebre @sbueringer @chrischdi

t-lo replied:
Added you and Stefan. Will add others as they drop by in the PR comments and provide feedback.

t-lo commented:
Thank you Johanan, Fabrizio, and Stefan for tuning in! This is immensely helpful.
I also love the fact that we're already iterating over design thoughts; this is exactly the momentum we were hoping for with the working group.

Made a few changes to the proposal; reworked the whole user story part to focus on goals instead of implementations, and rephrased the "problem statement" section a bit to not hint at a solution when describing the issue.

Added a new section on stability and compatibility - this really was the proverbial elephant in the room for me since in Flatcar, we put a massive (occasionally painful) focus on _never_ breaking user workloads _ever_ - kudos to Stefan for calling this out. I'll make sure to hold the working group proposals to an equally high standard.

I don't think we're quite there yet but we're definitely making progress. Ready for another round of feedback!

t-lo commented:
@sbueringer , @fabriziopandini what do you think? Could you give it another pass?

johananl commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,107 @@
+
+Currently, the kubeadm bootstrap provider implements two provisioning systems:
+cloud-init and ignition.
+As there is currently no abstraction of node configuration, both implementations
```
What do we mean by "node configuration"? Is it different than "provisioning"?
I so, maybe we should clarify what this term means since "configuration" is a very overloaded term.
If not, I suggest we use consistent terminology to avoid further confusion around this matter.

t-lo replied:
We could agree on "OS provisioning" and could define it as "Configuration of the operating system at provisioning time".

johananl commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,107 @@
+MicroK8s currently only supports cloud-init.
+
+Outside of ClusterAPI, both [cloud-init](https://github.com/canonical/cloud-init)
+and [ignition](https://github.com/coreos/ignition) provisioning is widely adopted
```
```suggestion
and [ignition](https://github.com/coreos/ignition) provisioning are widely adopted
```

johananl commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,107 @@
+SUSE Linux, Alma, Rocky, Fedora, and Red Hat Enterprise Linux, ignition is popular
+with special-purpose distributions like Fedora CoreOS / Red Hat CoreOS, SuSE MicroOS
+/ SLE Micro / Kalpa, and Flatcar Container Linux.
+It is likely that more provisioning systems exist; a clean separation and guidelines
+will make it easier to add support to ClusterAPI as well as to share implementations
+across bootstrap providers.
```
```suggestion
It is likely that more provisioning systems exist; a clean separation between bootstrap and provisioning
and guidelines on how each is developed will make it easier to add support to ClusterAPI as well as to
share implementations across bootstrap providers.
```

johananl commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,107 @@
+across bootstrap providers.
+
+
+**Compatibility and API guarantees**
```
Does this mean this work group can't work on any changes which may result in breaking changes? IMO this will severely limit the possible solutions we could come up with.

Maybe we should mention that backward-compatible changes are _preferable_ and that in case breaking changes are introduced, a well-documented migration path with a proper deprecation process should be put in place?

t-lo replied:
That's a good point, I amended the section accordingly.

johananl commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,107 @@
+
+## Scope
+
+1. Propose architectural improvements that ease integration of node configuration
```
Again, not clear on what "node configuration" refers to.

t-lo replied:
I mixed two concerns into one (well, two) words. I've amended 1. accordingly.

johananl commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,107 @@
+1. Propose architectural improvements that ease integration of node configuration
+     implementations.
+   Decouple generic bootstrapping from configuration language specific concerns.
+2. Deliver an example implementation in the kubeadm bootstrapper.
```
```suggestion
2. Deliver an example implementation in the kubeadm bootstrap provider.
```

Better to use the [official terminology](https://cluster-api.sigs.k8s.io/reference/glossary#bootstrap-provider) to avoid confusion. Ditto all occurrences below.

johananl reviewed (commented):
Thanks! Added a few comments.

fabriziopandini commented:
I will come back to this after code freeze next week, I need to focus on stuff to get merged + CI signal

chrischdi commented:
Also showing up, I will need some more time to read myself into all of this but my focus now is first the upcoming CAPI release!

elmiko reviewed (commented):
in general this makes sense to me, i think it would be nice to have a few more details in the proposal. i left a couple suggestions.

also, i'm just starting to read the comments here.

t-lo commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,107 @@
+on the above.
+I would like to be able to do this to deliver choice, variety, and innovation to
+users of my own distribution as well as to the larger ClusterAPI ecosystem, and
+to enable other distribution vendors to do the same.
```
That's a good point, I made this part more specific.

t-lo commented:
Makes sense @fabriziopandini , thank you for the feedback! Let's look into at least a kick-off meeting and a recorded architecture discussion to bootstrap the WG (pun intended). I'll bring it up in today's ClusterAPI meeting.

schrej commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,111 @@
+
+## Scope
+
+1. Propose architectural improvements that abstract system configuration with the goal of
+   reducing duplicate functionality in current implementations, easing development of new configuration
+   features, and simplifying integration of new configuration systems for node bootstrapping.
```
I'm also interested in this idea in general (primarily adding a comment to this spot to have a thread). We're currently using metal3 to do bare metal provisioning, and have started work on a new bare metal oriented provider, both of which have a similar problem that most others do not: deploying outside of a cloud platform that take over most of the basic configuration.

To apply custom configuration it's still most convenient to use configuration systems like cloud-init etc., as that makes it a lot easier for users to provide custom images. This leads to two problems that are related to the scope of this WG:
- While systems like cloud-init allow you to provide "vendor" configuration in addition to the user configuration, the options for the former are somewhat limited (don't know about Ignition). In our experience it was necessary to use the bootstrap provider in order to apply configuration files, e.g. for advanced networking setups. This breaks the separation between infrastructure and Kubernetes configuration, and gets even more complex and tangled up when mixing in variables provided on a per-machine basis (possible with metal3 for example).
- Bare metal providers (or providers for platforms that do not come with enough automation) need to implement explicit support for configuration systems in order to use them to apply their configuration. If the system doesn't support providing multiple configurations (e.g. the inconsistent vendor/user config in cloud-init), the infra provider needs to merge configuration, making it even more complicated.

Both of these could be improved by allowing to extend the abstract configuration before turning it into a system-specific format.

I'm not sure if this topic came up already (I haven't read all the comments here or on the google doc), but I think this could be a good opportunity to make some improvements in CAPI geared towards on-prem use.

cc @kashifest - in case you're interested from a metal3 perspective

t-lo replied:
This sounds great - and I second the sentiment of making this a focus item in the working group. I've added a "Maintainability, Backwards Compatibility, and Extensibility" track to the WG agenda: https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?usp=sharing and took the liberty of adding you as a point of contact for these concerns.

Will close this thread to ease review of this PR as aligned on in the [Nov 27 office hours](https://docs.google.com/document/d/1GgFbaYs-H6J5HSQ6a7n4aKpk0nDLE2hgG2NSOM9YIRw/edit?tab=t.0#heading=h.faoavlb0ym66).

t-lo replied:
(Oh and @schrej  I've added you to the WG proposal as a Stakeholder, please let me know if that works for you)

kashifest replied:
I totally agree with @schrej's point of view here

mcbenjemaa reviewed (commented):
👀 please add me to the WG

t-lo commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,107 @@
+
+We will join the [regular ClusterAPI meetings](https://github.com/flatcar-hub/cluster-api?tab=readme-ov-file#-community-discussion-contribution-and-support)
+and share updates on our work in the "Feature Groups" section, which is a regular part of
+the ClusterAPI meetings)
```
Will add a link to our [WG minutes doc](https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?tab=t.0#heading=h.ac1b03gmo5ph) and update the markdown file as soon as we aligned on a meeting cadence as aligned on in the [Nov 27 Capi Office Hours](https://docs.google.com/document/d/1GgFbaYs-H6J5HSQ6a7n4aKpk0nDLE2hgG2NSOM9YIRw/edit?tab=t.0#heading=h.faoavlb0ym66).

Will resolve this conversation to make the PR more reviewable.

kashifest commented:
I would be interested to join the working group if there's still scope

t-lo commented:
Closed most feedback comments and added tracks / comments to our [working group minutes doc](https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit).

Added @schrej , @mcbenjemaa , and @kashifest as stakeholders.

t-lo commented:
* **WG Agenda Doc**: https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit
* **Poll for finding a meeting slot:** https://nuudel.digitalcourage.de/rnuXpCSWouj627uQ

CC: @johananl @tormath1 @fabriziopandini @elmiko @schrej @mcbenjemaa @kashifest @eljohnson92

johananl reviewed (commented):
Thanks @t-lo. I went through the whole document again. It looks much better to me but we still have multiple places where we use vague terms such as "configuration system". I suggest we remain consistent with either "bootstrap" (which is already defined in the CAPI [glossary](https://cluster-api.sigs.k8s.io/reference/glossary#bootstrap) or "provisioning" (which I highly recommend we **add** to the glossary) in order to avoid further conflation of these two surprisingly confusing concerns :slightly_smiling_face:

schrej commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,117 @@
+
+## Scope
+
+1. Propose architectural improvements that abstract system configuration with the goal of
+   reducing duplicate functionality in current implementations, easing development of new configuration
+   features, and simplifying integration of new configuration systems for node bootstrapping.
```
How about wording this a bit more explicitly, to avoid some of the overloaded terms:
```suggestion
1. Propose architectural improvements to the Machine provisioning process that separate Kubernetes bootstrapping tools (kubeadm, microk8s) from generating OS-distribution-specific initialization configuration (cloud-init, ignition) to avoid duplicate implementations, ease development of new configuration features and simplify implementing new providers for both kinds of tools.
```

johananl replied:
@schrej I like this except "initialization configuration" because we already call the same thing "provisioning" elsewhere and I think we should try to have one term for each concern since this area of the project is confusing as it is.

schrej replied:
The term is also used to describe the entire process of bringing up a new Machine, including creating the infrastructure, configuring the OS and setting it up as a K8s node. That's why I tried avoiding it here.

t-lo replied:
It's not just (necessarily) OS / distribution specific initialisation - e.g. with Ignition one could perform significant customisation like setting the root partition filesystem, creating / adding partitions, using encrypted FS / encrypted volumes, and the like. Ignition runs in the initrd so supports significant customisation options.

I'd like to stick with "OS provisioning" for the time being, will update the section accordingly.

t-lo commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,117 @@
+
+Currently, the kubeadm bootstrap provider implements two provisioning systems:
+cloud-init and ignition.
+As there is currently no abstract way to pass configuration to  node bootstrapping,
```
That's a good point, we might as well work out a more precise glossary in the WG. It will come in handy down the road. I added a "glossary" track to our WG agenda doc: https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?tab=t.0#heading=h.ed54x9s57jp1

t-lo commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,117 @@
+Currently, the kubeadm bootstrap provider implements two provisioning systems:
+cloud-init and ignition.
+As there is currently no abstract way to pass configuration to  node bootstrapping,
+both configuration system implementations - ignition and cloud-init - must cover
```
I'll use "OS provisioning" to be more precise.

t-lo commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,117 @@
+both configuration system implementations - ignition and cloud-init - must cover
+generic configuration as well as config language specific concerns.
+This leads to duplicate code and to maintenance overhead.
+Also, other providers like e.g. MicroK8s must implement their own OS configuration
```
Same - I'll use "OS provisioning". Thanks for the pointer!

t-lo commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,117 @@
+generic configuration as well as config language specific concerns.
+This leads to duplicate code and to maintenance overhead.
+Also, other providers like e.g. MicroK8s must implement their own OS configuration
+generators since kubeadm's node bootstrapping implementations cannot be re-used.
```
OS provisioning indeed. Good catch!

t-lo commented:
Gave it another pass, addressed all comments. I'm consistently using "node bootstrapping" for the process of CAPI creating a node, and "OS provisioning" for customising an operating system to match the layout/requirements of a node.
Also added a "glossary" track to the WG concerns (in the agenda doc): https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?tab=t.0#heading=h.ed54x9s57jp1

johananl commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,118 @@
+
+**Problem statement / Example issues**
+
+Currently, the kubeadm bootstrap provider implements two systems to configure a node's
+operating system at provisoining time (referred to as "OS provisioning" from here on):
+cloud-init and ignition.
+As there is currently no abstract way to pass OS provisioning configuration to node
+bootstrapping, both OS provisioning system implementations - ignition and cloud-init - must cover
+both generic node bootstrapping as well as OS provisioning system specific concerns.
+This leads to duplicate code and to maintenance overhead.
+Also, other providers like e.g. MicroK8s must implement their own OS provisioning
+systems since kubeadm's implementations is specific to kubeadm and cannot be re-used.
+MicroK8s currently only supports cloud-init.
```
```suggestion
As there is currently no OS provisioning configuration abstraction, the kubeadm bootstrap provider
is tightly coupled with cloud-init and Ignition. Furthermore, the Ignition implementation is built
on top of the cloud-init one. This makes it hard to develop provisioning implementations
independently and makes it hard to reuse code effectively between bootstrap implementations: Other
bootstrap providers such as MicroK8s must implement their own OS provisioning, and MicroK8s
currently only supports cloud-init.

In addition, the current design mixes bootstrap-related provisioning code (such as the commands
necessary for executing the kubeadm binary on a host) with generic OS customizations (such as disk
partitioning or Linux user configuration).
```

I suggest this rephrasing since IMO the current phrasing doesn't emphasize the extent of the current issues.

Our main problems:

- Bootstrap is tightly coupled with Provisioning. This is the biggest issue IMO.
- The Ignition implementation is built on top of the cloud-init implementation. This means that we not only have a tight coupling between bootstrap and provisioning, **in addition** Ignition fully depends on cloud-init.

The implications of these problems:

- We can't modify cloud-init or Ignition code without affecting kubeadm-related code or vice versa.
- We can't modify cloud-init code without affecting Ignition code.
- We can't flexibly combine bootstrap providers with provisioning implementations (regardless of the question of whether provisioning should be embedded inside the bootstrap provider or be fully independent).
- Provisioning config is "abused" by users who want to perform generic OS customization (unrelated to CAPI bootstrap, i.e. disk partitioning) and have no well-defined mechanism for doing so.

I've explained all of the above in detail under the [Motivation section](https://docs.google.com/document/d/1Fz5vWwhWA-d25_QDqep0LWF6ae0DnTqd5-8k8N0vDDM/edit?tab=t.0#heading=h.wcf1gbhz0yw1) of the WIP design proposal (I think it's relevant to read that section even if we don't end up adopting the specific proposed design).

t-lo replied:
This is great! Thank you for elaborating.

johananl commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,118 @@
+1. Propose architectural improvements that abstract OS provisioning configuration with the goal of
+   reducing duplicate functionality in current implementations, easing development of new OS provisioning
+   features, and simplifying integration of new configuration systems for OS provisioning.
+2. Deliver an example implementation in the kubeadm bootstrap provider.
```
```suggestion
2. Deliver an example implementation which works with the kubeadm bootstrap provider.
```

The current phrasing implies that provisioning is embedded inside bootstrap -- which is one problem we are trying to address so I'm not sure we want to prescribe this assumption in advance. We may well decide that we want to keep things as they are, but let's at least discuss things first.

t-lo replied:
Good point; and I believe Fabrizio and you are in alignment with this.

elmiko commented:
i don't have any specific suggestions or recommendations at this time, but i think this is going in the right direction and i appreciate @johananl 's latest comments.

schrej commented:
/lgtm

t-lo commented:
Rebased to latest main, squashed to a single commit, and added the meeting cadence and link to the meeting calendar.

@schrej could you please have another look?

schrej reviewed (approved):
/lgtm

johananl commented:
/ok-to-test

johananl commented:
@t-lo the CI doesn't like the link to the calendar file. I wonder if it's OK to ignore this test.

rothgar commented:
I'm a bit confused by the problem statement in the document.

1. Is this intended to only apply to configuration _after_ the OS has booted and been installed? Cloud-init and ignition make a lot of assumptions about what's on the node and cannot be abstracted.
2. Is this only intended to be used for `kubeadm` based provisioning? I know it mentions other options but the focus and ultimate implementation is based on kubeadm with others being optional
3. Is this meant to re-configure nodes after they've been provisioned? From my understanding this is mostly designed to configure a node from a base, unconfigured state. Ignition and cloud-init don't re-configure or modify nodes and the only option is to destroy/replace

I'm trying to figure out if I should be involved at all (right now it seems like the answer is no) because Talos Linux doesn't use ignition, cloud-init, or kubeadm to configure (and re-configure) a node for kubernetes.

johananl commented:
Hey @rothgar.

This working group is about improving the separation between bootstrap and provisioning. "Bootstrap" means "turning a server into a k8s node". "Provisioning" means "running cloud-init/Ignition to customize a server". "Server" in this context means a physical bare metal server or a VM.

We don't intend to alter the **functionality** of bootstrap or provisioning in this working group, we're concerned mainly with how we can maintain and evolve multiple provisioner implementations (e.g. cloud-init, Ignition) over time. This does entail touching CABPK (the kubeadm bootstrap provider) because right now bootstrap and provisioning are tightly coupled, which means we can't touch e.g. cloud-init code without touching kubeadm-related code.

We hope to get to an acceptable state in CABPK which could serve as a "template" for other bootstrap providers, however there's still a way to go before we get there. We want to avoid making API changes to the extent that's possible, but of course if we do end up proposing API changes then this **would** affect all relevant providers, including Talos.

I suggest you keep an eye on the products of this working group to determine if/when you need to be involved, however I think that at the moment this work is irrelevant for anything which doesn't use kubeadm, cloud-init or Ignition. We'll be sure to broadcast any significant changes loudly to the community in case we realize they're inevitable.

chrischdi commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,121 @@
+
+* Meeting schedule: Fortnightly; every first and third Thursday of the month at 17:00 UTC
+  (WG [calendar](https://calendar.google.com/calendar/u/0?cid=OTBkMjJjZGU0OTcyZjI0OGQ2NTE2YTk2ZGUwNWVmNjI1NTM2NDRmZmYyNjFlMjE1MGY1ZjIyOTU0NmQ1OWQ0MUBncm91cC5jYWxlbmRhci5nb29nbGUuY29t),
+      [iCal](https://calendar.google.com/calendar/ical/90d22cde4972f248d6516a96de05ef62553644fff261e2150f5f229546d59d41%40group.calendar.google.com/public/basic.ics)).
```
```suggestion
      [iCal](https://calendar.google.com/calendar/ical/90d22cde4972f248d6516a96de05ef62553644fff261e2150f5f229546d59d41%40group.calendar.google.com/public/basic.ics).
```

t-lo commented:
I'd like to keep the iCal link - it's very convenient to be able to import the meeting series. Here's the link, it works: https://calendar.google.com/calendar/ical/90d22cde4972f248d6516a96de05ef62553644fff261e2150f5f229546d59d41%40group.calendar.google.com/public/basic.ics Tested on a number of machines. If we could figure out what makes this test fail then that would be awesome...

@rothgar If you like to know more, feel free to join our WG office hours Thursday next week (Dec 19), at 5pm UTC - happy to chat and to give an intro to the whole thing.

t-lo commented:
/ok-to-test

t-lo commented:
/retest

t-lo commented:
/retest

schrej commented:
/lgtm

t-lo commented:
@fabriziopandini @elmiko I think this can be merged now? The WG is well established, up and running. First stabs at implementations are in progress, and we also started related contributions to development docs (Johanan is working on a PR to dev docs on using KubeVirt as a dev environment).

elmiko reviewed (commented):
yes, i think so too @t-lo . thanks for the work here =)

/approve
/lgtm

johananl commented:
/assign sbueringer

sbueringer commented:
/lgtm

Would be great if other maintainers can also take a (final) look 

/assign @chrischdi @enxebre @fabriziopandini @vincepri

fabriziopandini commented:
/lgtm

sbueringer commented:
@t-lo I would recommend bringing this up one last time in the office hours and setting a lazy consensus of 1-2 weeks

t-lo commented:
> @t-lo I would recommend bringing this up one last time in the office hours and setting a lazy consensus of 1-2 weeks

I'll bring it up in today's call, thanks @sbueringer .

sbueringer commented:
As discussed in the office hours, lazy consensus until February 19

If any of the other maintainers want to chime in, now would be the time

enxebre commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,121 @@
+bootstrap providers.
+I would like to read and to follow documentation, guidelines, and specifications
+on the above.
+I would like to offer a choice of node bootstrapping configuration systems to users,
```
can we include an illustrative example? e.g "I would like to enable the bootstrap API consumer to express intent to fine tune nodes by setting vm.dirty_ratio while not having to reimplement the logic to setup kubelet and let the machine become a Node" if that's a good representative?

t-lo replied:
This is a great idea!
Probably the not a perfect example though, as setting sysctls solely involves the OS layer - i.e. it does not strictly fulfil a provisioning need of the bootstrap provider. IOW we can provision the nodes perfectly fine and alternatively set the sysctl later, e.g via an operator. Also, this opens a whole can of works as we would later need to validate the intent (and let the provisioning fail if it doesn't work) etc. 

How about something more fundamentally provisioning related like disks set-up or disk partitioning? E.g. adding a required storage device / file system to a mount point? I'll add something in the doc.

enxebre commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,121 @@
+
+**Problem statement / Example issues**
+
+As there is currently no OS provisioning configuration abstraction, the kubeadm bootstrap provider
```
it's probably worth mentioning as part of the context that our API also has a knob that enables full customization via custom userdata secret

enxebre commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,121 @@
+    - "@johananl"
+    - "@tormath1"
+    - "@fabriziopandini"
+    - "@sbueringer"
```
happy to be included as reviewer as well

enxebre commented:
I dropped some feedback, no blockers. lgtm.

mauromorales commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,121 @@
+Node Bootstrapping.
+
+Provisioning, in this context, refers to configuring and customising the
+node operating system to prepare it to serve as a ClusterAPI cluster node.
```
```suggestion
node's operating system to prepare it to serve as a ClusterAPI cluster node.
```

mauromorales commented on docs/community/20241112-node-bootstrapping.md:
```diff
@@ -0,0 +1,121 @@
+stakeholders for a Working Group dedicated to the Provisioning aspect of
+Node Bootstrapping.
+
+Provisioning, in this context, refers to configuring and customising the
```
Suggesting to change to American English just because later on is used. But the suggestion can be inverted if you prefer the British version.

```suggestion
Provisioning, in this context, refers to configuring and customizing the
```

fabriziopandini commented:
@t-lo if you can address last nits, we can get this merged

t-lo commented:
Updated the PR:
- Addressed @enxebre 's nits and added a concrete provisioning-time customization example. Added them to the lost of reviewers.
- Included @mauromorales' feedback and added them to the list of stakeholders.

fabriziopandini commented:
Lazy consensus is expired, and the WG is now the venue to address further comments or ideas.
Let's keep moving!

/lgtm
/approve
